contracts/Richman.sol

```diff
@@ -0,0 +1,167 @@
+pragma solidity ^0.4.15;
+
+contract Richman {
+    address private owner;
+    uint private totalSupply;
+    mapping(address => uint) private ledger;
+    mapping(address => bool) private blacklist;
+    
+    uint public freeTokens; //  tokens, available to borrow
+    
+    
+    //  Modifiers
+    modifier isOwner(address _addr) {
+        require(_addr == owner);
+        _;
+    }
+    
+    modifier notOwner(address _addr) {
+        require(_addr != owner);
+        _;
+    }
+    
+    /** @dev Checks if enough of free tokens to lent
+      * @param _amount Amount needed to check
+    */
+    modifier amountIsAvailable(uint _amount) {
+        require(freeTokens >= _amount);
+        _; 
+    }
+    
+    /** @dev Checks borrower has borrowed tokens
+      * @param _addr Address of borrower to check
+    */
+    modifier hasBorrowedTokens(address _addr) {
+      require(ledger[_addr] > 0);
+        _;
+    }
+    
+    modifier checkedForBorrowOverflow(address _addr, uint _amount) {
+        require( (freeTokens - _amount) < freeTokens &&
+                ledger[msg.sender] < (ledger[msg.sender] + _amount) );
+                
+        _;
+    }
+    
+     modifier checkedForPayBackUnderflow(address _addr, uint _amount) {
+        uint freeTokensWithPayBack = freeTokens + _amount;
+        uint borrowedAmount = ledger[msg.sender];
+
+        require( (freeTokensWithPayBack - freeTokens) == _amount &&
+                (borrowedAmount - _amount) < borrowedAmount );
+                
+        _;
+    }
+    
+    modifier moreThanZero(uint _amount) {
+        require(_amount > 0);
+        _;
+    }
+    
+    modifier notInBlacklist(address _addr) {
+        require(!blacklist[_addr]);
+        _;
+    }
+    
+    modifier isValidAddress(address _addr) {
+        require(_addr != 0x0);
+        _;
+    }
+    
+    //  Events
+    event LogBorrowSuccess(address _addr, uint _amount);
+    event LogPayBackSuccess(address _addr, uint _amount, string _str, uint _debt);
+    event LogTotalSupplyUpdateStatus(bool success);
+    event LogFallbackCalled(address _addr, string _str);
+    
+    
+    function Richman(uint _totalSupply) public payable {
+        owner = msg.sender;
+        
+        totalSupply = _totalSupply;
+        freeTokens = _totalSupply;
+    }
+    
+    function () payable public { 
```

Comment:
As far, as I know there can't be any logic in fallback functions because of gas max 2300. Firstly I was going to add address with amount into fallbackTokens mapping. But it is impossible, cause we can't do any of this: 
Writing to storage
Creating a contract
Calling an external function which consumes a large amount of gas
Sending Ether
So I decided to send event only. suicide() function can be added for some case, so all ethers will be moved to owner.

Reply:
1. 2300 is only available in address.send/transfer, in other cases it is not limited to some constant.
2. Why accepting ETH at all, if there is no use of it? Of there is no payable fallback then all ETH will be reverted by default.